apollo-federation/src/type_merging.rs

```diff
@@ -0,0 +1,1364 @@
+    ) -> Result<bool, FederationError>
+    where
+        TElement: SchemaElementWithType + 'a,
+    {
+        // Validate sources
+        if sources.is_empty() {
+            return Err(SingleFederationError::Internal {
+                message: "No type sources provided for merging".to_string(),
+            }
+            .into());
+        }
+
+        let mut result_type: Option<Type> = None;
+        let mut has_subtypes = false;
+        let mut has_incompatible = false;
+
+        // First pass: determine the merged type following GraphQL Federation variance rules
+        for (_index, source_type) in type_sources_helpers::iter_types(sources) {
+            let Some(source_type) = source_type else {
+                continue;
+            };
+
+            if let Some(ref current_type) = result_type {
+                if self.same_type(current_type, source_type) {
+                    // Types are identical, continue
+                    continue;
+                } else if let Ok(true) = self.is_strict_subtype(current_type, source_type) {
+                    // current_type is a subtype of source_type (source_type is more general)
+                    has_subtypes = true;
+                    if is_input_position {
+                        // input: use more general (supertype) → upgrade to source_type
+                        result_type = Some(source_type.clone());
+                    }
+                    // else output: keep current_type (the subtype)
+                } else if let Ok(true) = self.is_strict_subtype(source_type, current_type) {
+                    // source_type is a subtype of current_type (current_type is more general)
+                    has_subtypes = true;
+                    if !is_input_position {
+                        // For input types, use the more specific type (contravariance)
+                        result_type = Some(source_type.clone());
+                    }
+                    // else output: keep current_type (the supertype)
+                } else {
+                    // Types are incompatible
+                    has_incompatible = true;
+                }
+            } else {
+                // First type we encounter
+                result_type = Some(source_type.clone());
+            }
+        }
+
+        let Some(typ) = result_type else {
+            // No sources provided - this shouldn't happen in normal operation
+            return Err(FederationError::SingleFederationError(
+                SingleFederationError::Internal {
+                    message: format!(
+                        "No source provided for {} across subgraphs",
+                        dest.coordinate()
+                    ),
+                },
+            ));
+        };
+
+        // Copy the type reference to the destination schema
+        let copied_type = match self.copy_type_reference(&typ, self.schema) {
+            Ok(copied) => copied,
+            Err(_) => {
```

Comment:
I think this is wrong, if you get an error, I think it is right to propagate the error up rather than try to correct